src/Processor/HotspotProcessor.php

```diff
@@ -7,29 +7,120 @@
+            // Get point X and Y for crop image based on hotspot
+            [$offset_x, $offset_y] = $this->resolveCropOffset($image, $width, $height, $hotspot);
+
+            // Debug mode draw crop and center point on image
+            if ($this->debug) {
+                $center_x = (int) round($hotspot[1] * $image->width());
+                $center_y = (int) round($hotspot[2] * $image->height());
+
+                $x1 = (int) min($image->width() - $width, max(0, $center_x - ($width / 2)));
+                $y1 = (int) min($image->height() - $height, max(0, $center_y - ($height / 2)));
+
+                $x2 = (int) max($width, min($image->width(), $center_x + ($width / 2)));
+                $y2 = (int) max($height, min($image->height(), $center_y + ($height / 2)));
+
+                /**
+                 * Upgrade Intervention Image to 3.x
+                 * rectangle() is now handled by drawRectangle()
+                 * @see https://image.intervention.io/v3/modifying/drawing#drawing-a-rectangle
+                 */
+                // Draw rectangle on final crop
+                $image->rectangle($x1, $y1, $x2, $y2, function ($draw) {
+                    $draw->border(3, '#0000FF');
+                });
+
+                /**
+                 * Upgrade Intervention Image to 3.x
+                 * ellipse() is now handled by drawEllipse()
+                 * @see https://image.intervention.io/v3/modifying/drawing#drawing-ellipses
+                 */
+                // Draw green ellipse in center
+                $image->ellipse(30, 30, $center_x, $center_y, function ($draw) {
+                    $draw->border(3, '#0FF000');
+                });
+
+                return;
+            }
+            $image->crop($width, $height, $offset_x, $offset_y);
+        }
+    }
+
+    /**
+     * @param array{int, int} $crop
+     *
+     * @return array{int, int}
+     */
+    private function getWidthHeight(Image $image, array $crop): array
+    {
+        // Square ratio
+        if ($crop[0] == $crop[1]) {
+            $width = $height = min($image->width(), $image->height());
+        } elseif ($crop[0] > $crop[1]) { // Horizontal ratio
+            $width = $image->width();
+            $height = (int) round(($image->width() * $crop[1]) / $crop[0]);
+        } else { // Vertical ratio
+            $width = (int) round(($image->height() * $crop[0]) / $crop[1]);
+            $height = $image->height();
+        }
+
+        return [$width, $height];
+    }
 
-            $cropWidth = (int) $hotspot[3];
-            $cropHeight = (int) $hotspot[4];
+    /**
+     * @param array{string, float, float} $hotspot
+     *
+     * @return array{int, int}
+     */
+    private function resolveCropOffset(Image $image, int $width, int $height, array $hotspot): array
+    {
+        $offset_x = (int) (($image->width() * $hotspot[1]) - ($width / 2));
+        $offset_y = (int) (($image->height() * $hotspot[2]) - ($height / 2));
```

Comment:
Be careful, casting to `int` without `rounding` before is a truncate, so you may lose some precision,


```suggestion
        $offset_x = (int) round(($image->width() * $hotspot[1]) - ($width / 2));
        $offset_y = (int) round(($image->height() * $hotspot[2]) - ($height / 2));
```